### PR TITLE
Reserve memory for host

### PR DESCRIPTION
**Pull request by ravening**
## Description

By default cloudstack reserves 1Gb of RAM in hosts
using _dom0_memory field. Add a global setting
"host.reserved.mem.mb" which can used to either
increase or decrese the amount of memory which can be reserved

## Types of changes

- [X] Enhancement (improves an existing feature and functionality)

## How Has This Been Tested?

Total capacity and RAM of each host before setting the global setting

```
mysql> select total_capacity,used_capacity from op_host_capacity where capacity_type=0\G
*************************** 1. row ***************************
total_capacity: 7297130496
 used_capacity: 6174015488
*************************** 2. row ***************************
total_capacity: 7297130496
 used_capacity: 0
*************************** 3. row ***************************
total_capacity: 7297130496
 used_capacity: 0
3 rows in set (0.00 sec)

mysql> select dom0_memory,ram from host where type="Routing" and removed is null\G
*************************** 1. row ***************************
dom0_memory: 1073741824
        ram: 7297130496
*************************** 2. row ***************************
dom0_memory: 1073741824
        ram: 7297130496
*************************** 3. row ***************************
dom0_memory: 1073741824
        ram: 7297130496
3 rows in set (0.01 sec)
```

setting global setting `host.reserved.mem.mb` to 4096

```
mysql> select total_capacity,used_capacity from op_host_capacity where capacity_type=0\G
*************************** 1. row ***************************
total_capacity: 4075905024
 used_capacity: 6174015488
*************************** 2. row ***************************
total_capacity: 4075905024
 used_capacity: 0
*************************** 3. row ***************************
total_capacity: 4075905024
 used_capacity: 0
3 rows in set (0.00 sec)

mysql> select dom0_memory,ram from host where type="Routing" and removed is null\G
*************************** 1. row ***************************
dom0_memory: 4294967296
        ram: 4075905024
*************************** 2. row ***************************
dom0_memory: 4294967296
        ram: 4075905024
*************************** 3. row ***************************
dom0_memory: 4294967296
        ram: 4075905024
3 rows in set (0.00 sec)
```

Corresponding entry in `agent.properties`

`host.reserved.mem.mb=4096`

### CODE REVIEW

**Comment by shwstppr** [comment]
@ravening is this ready? can you please fix the conflict

**Comment by ravening** [comment]
> @ravening is this ready? can you please fix the conflict

updated pr

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@ravening this already exists for KVM via the setting `host.reserved.mem.mb` in agent.properties for per KVM/host basis. If you want to introduce a change, instead of adding a new setting, why not extend this and make this a cluster scoped global setting that is propagated and saved/enforced via LibvirtComputingResource and saved in each host's agent.properties file?

**Comment by ravening** [comment]
> host.reserved.mem.mb

@rhtyd I just checked the code and the global setting does not exist. The value `host.reserved.mem.mb` is only checked in `plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java` and has a default value of 1GB. This pr adds a global setting so that we can customize the value and the new value will be stored in all agents. The value is updated in the `host` table as well

**Comment by rhtyd** [comment]
@ravening if you read again, I said the setting `host.reserved.mem.mb` exists in `agent.properties` and read by the KVM agent (you're right the `LibvirtComputingResource`) and allows the setting to be configured on per host basis. If you want to pull this config in mgmt server, you can do so and define a new ConfigKey that is cluster scoped and changes are propagated and saved in agnet.properties file.

**Comment by ravening** [comment]
> @ravening if you read again, I said the setting `host.reserved.mem.mb` exists in `agent.properties` and read by the KVM agent (you're right the `LibvirtComputingResource`) and allows the setting to be configured on per host basis.
> 
> 
> 
> If you want to pull this config in mgmt server, you can do so and define a new ConfigKey that is cluster scoped and changes are propagated and saved in agnet.properties file.

@rhtyd yes I created the global setting with the same name and made it global but it's not cluster wide. Should it be cluster wide?

**Comment by shwstppr** [comment]
@ravening Conflicts here too. Moving to 4.16.0.0 milestone based on base branch please update if needed

**Comment by ravening** [comment]
> @ravening Conflicts here too. Moving to 4.16.0.0 milestone based on base branch please update if needed

@shwstppr resolved the conflicts

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
ping @ravening

**Comment by sureshanaparti** [comment]
Hi @ravening can you fix conflicts, and rebase to latest main.

**Comment by nvazquez** [comment]
Ping @ravening

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
not sure what happened , but both travis and blueorangutan fail to build @ravening

**Comment by rhtyd** [comment]
Ping @ravening

**Comment by ravening** [comment]
@rhtyd @sureshanaparti done

**Comment by nvazquez** [comment]
Hi @ravening there is a checkstyle error on packaging, can you please fix and test building for checking other failures?

````
[ERROR] /jenkins/workspace/acs-centos7-pkg-builder/dist/rpmbuild/BUILD/cloudstack-4.16.0.0-SNAPSHOT/server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:206:1: Duplicate import to line 20 - com.cloud.configuration.ConfigurationManagerImpl.SET_HOST_DOWN_TO_MAINTENANCE. [RedundantImport]
````

**Comment by weizhouapache** [comment]
@ravening can you remove line 20 - com.cloud.configuration.ConfigurationManagerImpl.SET_HOST_DOWN_TO_MAINTENANCE ?

**Comment by ravening** [comment]
@weizhouapache @nvazquez reorganized the imports

**Comment by sureshanaparti** [comment]
@ravening can you please address the outstanding comments, and fix the conflicts.

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 1]
```diff
@@ -1302,11 +1304,21 @@ public boolean configureHostParams(final Map<String, String> params) {
             storage.persist("router.aggregation.command.each.timeout", String.valueOf(longValue));
         }
 
+<<<<<<< HEAD
```
Comment (weizhouapache) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:1307]: @ravening 
this needs to be fixed :-D

Comment (ravening) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: @weizhouapache fixed it :P

**Comment by DaanHoogland** [comment]
again a conflict @ravening :(

**Comment by ravening** [comment]
> again a conflict @ravening :(



> again a conflict @ravening :(

@DaanHoogland i will wait for few more days before resolving conflicts... i have resolved conflicts several times but i still get this conflicts.

**Comment by DaanHoogland** [comment]
@ravening did you look at the other code? is it aligned with yours? conflicts are a sometimes sign that one of the functionalities should go in another module.

**Comment by ravening** [comment]
> @ravening did you look at the other code? is it aligned with yours? conflicts are a sometimes sign that one of the functionalities should go in another module.

@DaanHoogland resolved the conflicts

**Comment by nvazquez** [comment]
Hi @ravening can you please fix the conflicts?

**Comment by rohityadavcloud** [comment]
Sorry dhslove - my bad, the bot is picking wrong PR author, it was meant for @ravening

**Comment by ravening** [comment]
> Sorry dhslove - my bad, the bot is picking wrong PR author, it was meant for @ravening

@rohityadavcloud fixed it

**Pull request closed without merging**